---
title: "Applecorn run"
output:
  html_document:
    df_print: paged
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    number_sections: false
    theme: sandstone
    highlight: espresso
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE)
```

## Initial processing, DADA2 + phyloseq

Initial processing of amplicon data done using DADA2 and phyloseq packages. Briefly raw fastq files
quality filtered and trimmed, R1 and R2 reads overlaped to form a longer, merged reads and PCR duplicates
collapsed. Usually these steps collectivelly called denoising the data.

### DADA2

Initial step of getting denoised sequences table

```{r init}
library(drake)
loadd(dada2_res)
loadd(taxtab)
loadd(tree)
loadd(ps)
loadd(ps_filt)
loadd(alpha)
loadd(rare_curve)
```

```{r dada2_res1}
dada2_res[["seqtab_nochim"]] %>% dim
```

Checking that data is okay, typically expect very low chimeric rate

```{r dada2_res2}
dada2_res[["chim_rate"]]
```

Per sample break down of total number of denoised reads. This is the library size per sample

```{r dada2_res3
dada2_res[["info"]] %>% arrange(-nonchim)
```

### Donig taxonomy assignment

Assigning using green genes classifier. Total number of sequence and samples

```{r taxa_res1}
taxtab %>% dim
```

### Generating phyloseq object

phyloseq object

```{r phyloseq_obj}
ps
```

### Phylogeny tree

Plotting phylogeny tree to have a look how our ASVs behaving. This should be a routine step essentially at the start of any analysis. In this case we can see that there is one particular ASV that evidently doesn't fit the tree all that well. We should filter it out. This is particular important for downstream UniFrac analysis, since distance is calculated based on phylogeny as well as sequence similarity.

This is how the phylogenic tree looks like after filtering "an odd ball out"

```{r phylo_tree}
ps_filt[["raw_tree"]]
ps_filt[["filt_tree"]]
```

Based on the tree phyloseq object got filtered

```{r phyloseq_obj_filt}
ps_filt[["ps_filt"]]
```

Best practice as well would be BLAST that (those if multiple found) sequence to make sure it doesn't belong to your analysis. In this case this sequences BLASTed against mouse genome.

## Alpha diversity

Plot alpha diversity after rarefied to minimum depth

```{r, min_depth}
alpha[["min_depth"]]
```

```{r, alpha1}
alpha[["plot"]]
```

## Rarefaction curve

```{r rare_curve1, warning=F}
rare_curve
```
