---
title: "16S analysis"
output:
  html_document:
    df_print: paged
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    number_sections: false
    theme: sandstone
    highlight: espresso
-params:
-  origin: "."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```

## Initiation

Below is a reflection of the modular setup of the analysis. The module names
reflect the type of analysis / data manipulation that happens at that stage.

```{r cache=FALSE, include = F}

libs <- list.files(file.path(params$origin, "R"), full.names = TRUE)
lapply(libs, knitr::read_chunk)
```

Loading required libraries for the whole analysis

```{r libraries}
```

## Initial processing, DADA2 + phyloseq

Initial processing of amplicon data done using DADA2 and phyloseq packages. Briefly raw fastq files
quality filtered and trimmed, R1 and R2 reads overlaped to form a longer, merged reads and PCR duplicates
collapsed. Usually these steps collectivelly called denoising the data.

### DADA2

Initial step of getting denoised sequences table

```{r get_seqtab, echo = F}
```

Checking that data is okay, typically expect very low chimeric rate

```{r sanity_check1.1}
```

Per sample break down of total number of denoised reads. This is the library size per sample

```{r lib_size_info1}
```

### Donig taxonomy assignment

Assigning using green genes classifier

```{r taxonomy_assignment}
```

### Generating phyloseq object

```{r metadata}
```

phyloseq object

```{r phyloseq_obj}
```

### Phylogeny tree

Plotting phylogeny tree to have a look how our ASVs behaving. This should be a routine step essentially at the start of any analysis. In this case we can see that there is one particular ASV that evidently doesn't fit the tree all that well. We should filter it out. This is particular important for downstream UniFrac analysis, since distance is calculated based on phylogeny as well as sequence similarity.

This is how the phylogenic tree looks like after filtering "an odd ball out"

```{r phylo_tree}
```

Best practice as well would be BLAST that (those if multiple found) sequence to make sure it doesn't belong to your analysis. In this case this sequences BLASTed against mouse genome.

```{r phylo_tree_filt}
```

## Rarefaction curve

```{r, rare_curve1, warning=F}
```

```{r, rare_curve2, warning=F}
```

## Alpha diversity

Plot alpha diversity after rarefied to minimum depth

```{r, alpha1, fig.width = 10, fig.height=6, warning=F}
```

## Ordination

```{r ordination_init, warning=FALSE}
```

### Presence and absence

```{r uw_unifrac_plot}
```

### Abundance based

It is good idea to compare Bray-Curtis dissimilarity distance with weighted UniFrac, since UniFrac takes phylogenetic relationship into consideration. The plot should resemble the same overall shape of the data. If the two plots don't appear similar then it is possible that there are some sequence contaminats, that causing outlieers in the phylogenetic tree. Those need to be removed, before re-calculating weighted UniFrac distance.

```{r bray_plt}
```

```{r w_unifrac_plot1}
```

```{r mk_ord_plots, echo = F}
```

## PERMANOVA

This is to set out significance that we are seeing on the weighted UniFrac PCoA plot about

```{r permanova_pairwise, warning = F}
```

## Taxonomy

```{r main_df_setup}
```

```{r taxonomy}
```

- [barplot link](images/taxa.html)

## Differential taxa occurance

```{r deseq2}
```

```{r viz_volcano}
```

- [volcano plot link](images/volcano.html)

```{r viz_ma}
```

- [ma plot link](images/ma.html)
