---
title: "16S analysis"
output:
  html_document:
    df_print: paged
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    number_sections: false
    theme: sandstone
    highlight: espresso
params:
  origin: "."
  dada: NA
  taxtab: NA
  ps: NA
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```

```{r libraries}
```

## High level overview of typical amplicon data analysis

- FASTQ quality filtering (maybe will need to demultiplex before that)
- primers removal: cutadapt or "naive" length(primer) cut
- merging R1 and R2 reads into single targeted fragment
- Dereplication - collapsing identical sequences
- OTUs/ASVs picking (operational taxonomical unit / amplicon sequence variant)
- OTUs/ASVs annotation
- explorative analysis - alpha/beta diversities, various ordinations plots and taxonomic barplot, all to see what the data looks like and if it fit the biology
- differential taxa occurrance
- More? PICRUSt, build your own linear model

## Initial processing, DADA2 + phyloseq

Initial processing of amplicon data done using DADA2 and phyloseq packages. Briefly raw fastq files
quality filtered and trimmed, R1 and R2 reads overlaped to form a longer, merged reads and PCR duplicates
collapsed. Usually these steps collectivelly called denoising the data.

### DADA2

```{r echo = T}
dada2 <- params$dada
seqtab <- dada2[["seqtab"]]
seqtab_nochim <- dada2[["seqtab_nochim"]]
chim_rate <- dada2[["chim_rate"]]
info <- dada2[["info"]]

seq_info <- seqtab %>% dim
seq_nochim_info <- seqtab_nochim %>% dim
```

Total number of samples is `r seq_info[1]` with total number of unique ASVs called is `r seq_info[2]`. After removal of chimeric reads, total number of unique ASVs is `r seq_nochim_info[2]`. Chimeric rate is `r chim_rate`

### Denoising information

```{r, rows.print = 20}
info %>% arrange(-nonchim)
```


