---
title: "Applecorn run"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    number_sections: false
    theme: sandstone
    highlight: espresso
params:
  taxa_db: "DATABASE_PLACE_HOLDER"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE)
```

## Initialising the pipeline run

```{r init}
library(drake)
loadd(dada2_res)
loadd(taxtab)
loadd(tree)
loadd(ps)
loadd(ps_filt)
loadd(alpha)
loadd(rare_curve)
loadd(unifrac)
loadd(wunifrac)
```

## OTU picking and taxonomic assignment

This pipeline uses [DADA2 R package](https://benjjneb.github.io/dada2/index.html) for denosing the data and OTU picking. Denosing typically includes these few things:

- quality filtering and trimming of fastq files
- R1 and R2 reads merging into single, long read (amplicon)
- collapsing (filtering out) of PCR duplicates

[DADA2](https://benjjneb.github.io/dada2/index.html) also helps with OUT assignment against know database. In this run we used ``r params$taxa_db`` database for OTU assignment. Typically after [DADA2](https://benjjneb.github.io/dada2/index.html) processing a [phyloseq object](https://joey711.github.io/phyloseq/index.html) gets formed with the help of [phyloseq R package](https://joey711.github.io/phyloseq/index.html). This is a convenience package that offers a suite of function to analysis the amplicon data.


### DADA2 processing metrics

```{r dada2_res1}
d <- dada2_res[["seqtab_nochim"]] %>% dim
chim <- dada2_res[["chim_rate"]]
taxa_d <- taxtab %>% dim
```

```{r dada2_res2, include = F}
chk <- "low"
if(chim > 0.2) {
 chk <- "high"
}
```

Total number of samples in `r d[1]` with total number of OTUs `r d[2]`. Chimeric rate was `r chim`, which is `r chk`. Typically you expect very low chimeric rate. Number of OTUs remained after taxonomic assignment. It is possible that some OTUs weren't assigned to any class.

Table below provides per sample information of total number of inpurt raw reads and the reduction that happens as we move through denoising process.

```{r dada2_res3}
dada2_res[["info"]] %>% arrange(-nonchim)
```

### Generating phyloseq object

Typically four things go into making phyloseq object:

- table with denoised OTUs per sample
- experimental metadata
- table with assigned OTUs to class
- phylogenetic tree (optional, recommended)

```{r phyloseq_obj, comment = ""}
ps
```

## Phylogeny tree

Plotting phylogeny tree to have a look how our ASVs behaving. This should be a routine step essentially at the start of any analysis. In this case we can see that there is one particular ASV that evidently doesn't fit the tree all that well. We should filter it out. This is particular important for downstream UniFrac analysis, since distance is calculated based on phylogeny as well as sequence similarity.

```{r phylo_tree1}
ps_filt[["raw_tree"]]
```

This is phylogenic tree after filtering an outlier branches

```{r phylo_tree2}
ps_filt[["filt_tree"]]
```

We further need to filter those outlier sequences from the downstream analysis by filtering phyloseq object.

```{r phyloseq_obj_filt, comment = ""}
ps_filt[["ps_filt"]]
```

Generally it is good idea to also BLAST those outlier sequences to reassure yourself that they don't belong to the analysis

```{r phyloseq_obj_filt2, comment = ""}
seqs <- ps_filt[["outlier_seqs"]]
for(i in 1:length(seqs)) {
  n <- paste0(">Otu_", i)
  seq <- seqs[i]
  cat(n, seq, sep = "\n")
}
```

## Alpha diversity

```{r min_depth}
d <- alpha[["min_depth"]]
```

Plot alpha diversity after rarefied to minimum depth of `r d[1]`

```{r alpha1, warning = F}
alpha[["plot"]]
```

## Rarefaction curve

```{r rare_curve1, warning=F}
rare_curve
```

## Ordination

```{r ord1}
unifrac[["ord_plt"]]
wunifrac[["ord_plt"]]
```

## Session information

```{r sess_info, comment = ""}
sessionInfo()
```
